core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java

```diff
@@ -96,20 +105,34 @@ private void calculateData(IItemCollection input) {
 			IAggregator<?, ?> aggregator = Aggregators.distinct(query.getGroupBy());
 			Set<?> aggregate = input.getAggregate((IAggregator<Set<?>, ?>) aggregator);
 			if (aggregate != null) {
-				for (Object o : aggregate) {
-					IItemCollection rowCollection = input.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
-					Object[] row = newRow();
-					int column = 0;
-					for (; column < attributes.size(); column++) {
-						// Optimization - it is too expensive to do aggregation for these. You simply
-						// get first non-null matching attribute - we're only using this for the group by today.
-						row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+				int SUM_NUM_THREADS = Runtime.getRuntime().availableProcessors();
+				ExecutorService exec = Executors.newFixedThreadPool(SUM_NUM_THREADS);
+				try {
+					for (final Object o : aggregate) {
+						exec.submit(new Runnable() {
+							@Override
+							public void run() {
+								IItemCollection rowCollection = newInput
+										.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
+								Object[] row = newRow();
+								int column = 0;
+								for (; column < attributes.size(); column++) {
+									// Optimization - it is too expensive to do aggregation for these. You simply
+									// get first non-null matching attribute - we're only using this for the group by today.
+									row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+								}
+								for (int j = 0; j < aggregators.size(); j++) {
+									row[column + j] = rowCollection.getAggregate(aggregators.get(j));
+								}
+								data.add(row);
+							}
+						});
 					}
-					for (int j = 0; j < aggregators.size(); j++) {
-						row[column + j] = rowCollection.getAggregate(aggregators.get(j));
-					}
-					data.add(row);
+				} finally {
+					exec.shutdown();
 				}
+				// Timeout can be configured in prefrences (default value is set to 5 minutes)
+				exec.awaitTermination(ClassLeakingRule.CONFIGURED_TIMEOUT, TimeUnit.MINUTES);
```

Comment:
This means that the generic DefaultIItemResultSet class has an explicit dependency on the configuration of one rule, much better to add a timeout parameter at the creation of the instance.